### PR TITLE
Refactor command application for more consistency

### PR DESCRIPTION
**Pull request by james-j-obrien**
# Objective

- Prevent the case where a hook/observer is triggered but the source entity/component no longer exists 

## Solution

- Re-order command application such that all hooks/observers that are notified will run before any have a chance to invalidate the result.

## Testing
Updated relevant tests in `bevy_ecs`, all other tests pass.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
I agree with this change: it's a lot less of a footgun. The fact that this is simpler is nice too.

**Review by Diddykonga — APPROVED** [review]
LGTM, first in first out.

**Review by MiniaczQ — APPROVED** [review]
So much cleaner now 😄

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 1]
```diff
@@ -120,7 +120,21 @@ impl CommandQueue {
...
+        // Apply any commands in the world's internal queue first.
+        while !world.command_queue.is_empty() {
+            std::mem::swap(self, &mut world.command_queue);
+            self.apply_or_drop_queued(Some(world));
+        }
+
+        // Return the larger buffer to the world to be re-used
+        if self.bytes.capacity() > world.command_queue.bytes.capacity() {
+            std::mem::swap(self, &mut world.command_queue);
+        }
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/world/command_queue.rs:137]: Wondering if this juggling is necessary -- it makes the control flow a bit hard to follow and I'm still not 100% sure if every buffer will get put back in the right place.

Could we double-buffer the world's command queues? There could be a separate queue stored stored in the world as a private field, which we'd swap with the main queue when applying commands to reuse allocations. That should also prevent the capacity of unrelated system command queues from being inflated by commands added to the world's queue

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:137]: Yes, double buffering is a suitable alternative and I can swap to that instead.

Comment (JoJoJet) [reply crates/bevy_ecs/src/world/command_queue.rs:137]: I would prefer that. It seems simpler and more memory-efficient. It might end up making the code in this loop longer but I imagine it would compile to similar code after being unrolled

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs]: Fixed.

I wouldn't say it alleviates the juggling since you can't apply the additional buffer while it's still in the world, but it is better for ensuring buffers don't grow unnecessarily.

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 2]
```diff
@@ -120,7 +120,25 @@ impl CommandQueue {
         // flush the previously queued entities
         world.flush_entities();
 
-        self.apply_or_drop_queued(Some(world));
+        // Append our commands to the swap queue
+        world.swap_queue.append(self);
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/world/command_queue.rs:124]: Can we avoid this call to `append`? It will cause unnecessary allocations

I imagine that in most bevy apps, the world's command queue will be quite small, since it only stores commands enqueued by observers or hooks, whereas the size of system command queues will vary a lot. If there's a single system queues up a bunch of commands, this will force the size of the world's command queue to match that even when it's unnecessary

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs]: Fair point, fixed.

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 3]
```diff
@@ -120,7 +120,32 @@ impl CommandQueue {
...
+            std::mem::swap(self, &mut world.swap_queue);
+        }
+
+        // Apply our queue
         self.apply_or_drop_queued(Some(world));
+
+        // Apply any resulting commands from the world's internal queue
+        while !world.command_queue.is_empty() {
+            // Put our empty queue into the `swap_queue`
+            std::mem::swap(self, &mut world.swap_queue);
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/world/command_queue.rs:141]: Functionally this seems correct -- however the `std::mem::swap` between `self` and `world.swap_queue` seems unnecessarily ""clever"". Can you just have a loop that applies the world's command queue, then a loop that applies `self`'s command queue, followed by a loop that applies the world's command queue again if applicable?

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:141]: How do we apply the world's command queue without moving it out of the world?

We could create another command queue on the stack and juggle with that one instead but I don't think that makes anything simpler. Alternatively we can separate out a function that operates on the pointers and pass the pointers to the swap_queue using unsafe without removing it from world but that seems unnecessarily obtuse.

My first instinct was similar to yours but I think this may actually be the simplest implementation from where we stand now.

Comment (JoJoJet) [reply crates/bevy_ecs/src/world/command_queue.rs:141]: Hmm yeah. I suppose if any commands apply a `CommandQueue` of their own, then the world's `swap_queue` would have to get used.

**Review by JoJoJet — APPROVED** [review]

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 4]
```diff
@@ -120,7 +120,32 @@ impl CommandQueue {
...
         world.flush_entities();
 
+        // Apply any commands in the world's internal queue first.
+        while !world.command_queue.is_empty() {
+            // Put our queue into the `swap_queue`
+            std::mem::swap(self, &mut world.swap_queue);
+            // Take the full queue from the world and replace it with the `swap_queue`
+            std::mem::swap(self, &mut world.command_queue);
+            // Apply the queue
+            self.apply_or_drop_queued(Some(world));
```
Comment (cart) [inline crates/bevy_ecs/src/world/command_queue.rs:130]: Hmm have we considered the implications of commands applying CommandQueues inside of this "already swapped" context?

Quick pseudo code / inline memory representation:

```rust
world.swap_queue = SWAP[]
world.command_queue = WORLD[A]
self = LOCAL[C]

A {
   // applies a new command queue (other_queue = OTHER[E]), which queues new world command D
}

self.apply() {
    while !world.command_queue.is_empty() {
        world.swap_queue = LOCAL[C]
        self = WORLD[A, B]
        world.command_queue = SWAP[]
        self.apply_or_drop_queued() {
            Apply(A) {
                other_queue.apply() {
                    // no-op because SWAP is empty (the current value of world.command_queue)
                    while !world.command_queue.is_empty() {
                    }

                    Apply(E) {
                        // Applying E inserts D into world.command_queue, which is currently SWAP
                        world.command_queue = SWAP[D]
                    }
                    
                    while !world.command_queue.is_empty() {
                        world.swap_queue = OTHER[]
                        other = SWAP[D]
                        world.command_queue = LOCAL[C] 
                        other.apply_or_drop_queued() {
                            // the world.command_queue is now LOCAL[C] for this apply
                            // C may be applied here
                        }
                    }
                }
            }
        }
        // WORLD is now empty and "ready" to be the new SWAP queue
        world.swap_queue = WORLD[]
        self = LOCAL[] // now empty (see comment above)
        world.command_queue = SWAP[]
        self.apply_or_drop_queued() {
            // The code was written as if C would be applied here, but it was actually applied above a couple of nestings deep, and may have been _followed_ by other commands created during 
        }
    }
}
```

For the "world.command_queue is now `LOCAL[C]` for this apply" comment above, doesn't this have pretty serious order of operations implications?

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: Yeah that's not great, in that case we can just have an empty command queue on the stack that we move into `swap_queue` while we apply commands rather than inserting `self`.

Comment (cart) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: Might also be worth considering a sprinkle of unsafe to remove the swapping entirely. Both to avoid the brain breaking "nested swaps across contexts" situation and to optimize out a bunch of "unnecessary" swaps. Also, if we're creating a new empty queue on the stack, that also means we're allocating a bunch of extra times (and throwing it out on each call) instead of reusing already allocated space.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: We would only be allocating in the extra queue in the nested swap scenario, as long as the command queue application isn't recursive you would only be using the main two queues.

How do you picture unsafe being used here? We can't have the memory backing the command queue still in the world in any way since we could always end up writing to that queue due to some later nested command application and invalidating the memory we are reading.

Comment (cart) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: Given that we can read CommandMeta onto the stack, which is used to read the Command onto the stack, it seems like in theory we could ensure that there are no live references to the CommandQueue byte array while the command is being executed. If true, then provided we always look up directly using Vec indexing, then appending to the queue during command execution should be safe?

We'd also need to ensure we don't hold &mut CommandQueue over the command execution boundary.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: Yeah that's fair, it's a bit more of an involved refactor but I'll take a pass at it.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:130]: So this is tricky in the case that a command panics. Currently we have nice behaviour where we set the length of the vector before starting the loop to ensure it's not re-used. We can't do that if we want to allow appending to partially applied queues since the vector still needs it's state, so we need to have a cursor stored in the queue that we increment before hand to ensure we don't re-drop/apply the panic-ed command.

If we are okay with the additional `usize` in `CommandQueue` we should be good.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs]: Ok, this doesn't really work at all if we want MIRI to be happy. If we keep any pointers to the command queue and later someone uses the world to get a mutable reference to the command queue MIRI complains. I tried boxing the command queue, but it doesn't matter, it always violates stacked borrow rules.

@cart what's your suggestion here?

Comment (cart) [reply crates/bevy_ecs/src/world/command_queue.rs]: Spent a couple hours investigating this because I was pretty sure it should be possible. The solution I found includes the following:

1. Don't hold the mutable World borrow the whole time. Only turn it into a borrow at the exact point that it is used.
2. Box the World's CommandQueue. This decouples the memory from being stored directly on World, which ensures that `bytes` and `cursor` aren't "on the stack" when World is borrowed.
3. The addition of `RawCommandQueue`, which contains the raw pointers to `queue` and `cursor`. In addition to making some things a bit cleaner, it also allows us to use it in Commands (which was reworked to use an enum containing either `RawCommandQueue` or `Deferred<CommandQueue>`). This allows us to construct Commands from DeferredWorld without directly holding a mutable borrow to the CommandQueue.  I'm not convinced we actually need Commands to have the enum / branching. We might be able to have two separate types (one for `RawCommandQueue` and one for `Deferred<CommandQueue>`).

The branch is here: <https://github.com/cart/bevy/tree/raw_command_queue>. Needs a quality pass, a review of which new functions need to be unsafe, safety docs, and some design thinking about how RawCommandQueue should relate to Commands. I don't love the branching stuff.

I've gotta move on to other things for now. Let me know what you think / feel free to use whatever you want from that branch here.

Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs]: Definitely feels like we can get rid of the branching, but transitioning to only storing pointers should get around the stacked borrow problems MIRI was complaining about so this is closer to a solution than I am at the moment. 

Will merge this in, let people take a look, and then take another pass at clean-up when I have a chance.

**Review by hymm — COMMENTED** [review]
This is not a full review. Just did a quick pass over the unsafety.

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 5]
```diff
@@ -55,6 +61,58 @@ unsafe impl Send for CommandQueue {}
...
+        self.bytes.append(&mut other.bytes);
+    }
+
+    /// Returns false if there are any commands in the queue
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.cursor >= self.bytes.len()
+    }
+
+    pub unsafe fn get_raw(&mut self) -> RawCommandQueue {
```
[the remark this answers is not in the document]
Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs]: This is unsafe to bubble up the safety requirements you mention in your other comment, the invariant for the struct is it must always point to valid data, that way pushing to it is always valid. Since this is all (now) private to the crate their shouldn't be any way for users to violate this.

**Review by hymm — COMMENTED** [review]
made a pr to this branch clean up the safety invariants. This doesn't fix the memory leak issue.

https://github.com/james-j-obrien/bevy/pull/4

File: crates/bevy_ecs/src/world/command_queue.rs [hunk 6]
```diff
@@ -55,6 +70,85 @@ unsafe impl Send for CommandQueue {}
...
+    }
+}
+
+impl RawCommandQueue {
+    /// Returns a new `RawCommandQueue` instance.
+    pub fn new() -> Self {
+        // SAFETY: Pointers returned by `Box::into_raw` are guaranteed to be non null
+        unsafe {
+            Self {
+                bytes: NonNull::new_unchecked(Box::into_raw(Box::default())),
```
[the remark this answers is not in the document]
Comment (james-j-obrien) [reply crates/bevy_ecs/src/world/command_queue.rs:128]: This is handled for the one case it's actually constructed by the world `Drop` impl.
We could move the allocation into the world to make it more clear where the responsibility lies as any call to this constructor "needs" to free it.

**Review by hymm — APPROVED** [review]
The safety of the apis feels pretty clear now. You just need to guarantee that the raw command queue that the `Commands` is made from outlives it. And Miri is again happy with the cursed internals.
